models/transfer_check.go

```diff
@@ -175,6 +190,103 @@ func (t TransferDataCreateBody) ToIngestionMap(mapping TransferMapping) map[stri
 	}
 }
 
+func (t TransferDataCreateBody) FormatAndValidate() (TransferDataCreateBody, error) {
+	var err error
+	errs := make(FieldValidationError, 10)
+
+	// hash the iban if it's clear - otherwise keep it unchanged
+	t.BeneficiaryIban, err = hashIbanIfClear(t.BeneficiaryIban)
+	if err != nil {
+		errs["beneficiary_iban"] = err.Error()
+	}
+
+	// first validate the fields that expect a specific format
+	t.BeneficiaryBic, err = formatAndValidateBic(t.BeneficiaryBic)
+	if err != nil {
+		errs["beneficiary_bic"] = err.Error()
+	}
+
+	t.SenderBic, err = formatAndValidateBic(t.SenderBic)
+	if err != nil {
+		errs["sender_bic"] = err.Error()
+	}
+
+	t.Currency = strings.ToUpper(t.Currency)
+	if !slices.Contains(pure_utils.CurrencyCodes, t.Currency) {
+		errs["currency"] = fmt.Sprintf("currency %s is not valid", t.Currency)
+	}
+
+	_, err = netip.ParseAddr(t.SenderIP)
+	if t.SenderIP != "" && err != nil {
+		errs["sender_ip"] = fmt.Sprintf("sender_ip %s is not a valid IP address", t.SenderIP)
+	}
+
+	if !slices.Contains(TransferStatuses, t.Status) {
+		errs["status"] = fmt.Sprintf("status %s is not valid", t.Status)
+	}
+
+	stringTooLongErr := "string is too long"
+	// max length checks for strings that don't have any other format
+	if len(t.BeneficiaryName) > maxStringLengthTfCheck {
+		errs["beneficiary_name"] = stringTooLongErr
+	}
+	if len(t.Label) > maxStringLengthTfCheck {
+		errs["label"] = stringTooLongErr
+	}
+	if len(t.SenderAccountId) > idMaxLength {
+		errs["sender_account_id"] = stringTooLongErr
+	}
+	if len(t.TransferId) > idMaxLength {
+		errs["transfer_id"] = stringTooLongErr
+	}
+	if len(t.SenderDevice) > maxStringLengthTfCheck {
+		errs["sender_device"] = stringTooLongErr
+	}
+	if t.Value <= 0 {
+		errs["value"] = "value must be positive"
+	}
+
+	return t, errs
+}
+
+func formatAndValidateBic(bic string) (string, error) {
+	bic = strings.ToUpper(strings.TrimSpace(bic))
+	if len(bic) != 8 && len(bic) != 11 {
+		return "", errors.Wrap(BadParameterError, "bic must be 8 or 11 characters")
+	}
+	bicIsAlphanumeric := isAlphanumeric.MatchString(bic)
+	if !bicIsAlphanumeric {
+		return "", errors.Wrap(BadParameterError, "bic must be alphanumeric")
+	}
+
+	return bic[:8], nil
+}
+
+// takes a clear iban or a hexadecimal hash of an iban and returns a hexadecimal hash of the iban
+func hashIbanIfClear(ibanOrHash string) (string, error) {
+	data, err := hex.DecodeString(ibanOrHash)
+	if err == nil {
+		if len(data) == sha256.Size {
+			return trimAndUpper(ibanOrHash), nil
+		}
+	}
+
+	iban := trimAndUpper(ibanOrHash)
+	if !isAlphanumeric.MatchString(iban) {
+		return "", errors.Wrap(BadParameterError, "iban must be alphanumeric")
+	}
+	if len(iban) < 15 || len(iban) > 34 {
+		return "", errors.Wrap(BadParameterError, "iban must be between 16 and 34 characters")
+	}
+
+	hash := sha256.Sum256([]byte(trimAndUpper(iban)))
+	return strings.ToUpper(hex.EncodeToString(hash[:])), nil
+}
```

Comment:
IIUC, the spec is :
- compute the sha256 HASH
- encode in human readable hex format
- transform to UPPPERCASE since the format is case insensitive (to ensure we can easilly compare similar hash by string equality)

Reply:
Yes, with an additional step (first) to remove spaces and upper case (so that the hash is also comparable).
Actually, I'll change it a bit to "remove spaces" rather than simply "trim spaces" on the raw iban.